### PR TITLE
speed up tailing log for json-file log driver

### PR DESCRIPTION
**Pull request by haveatry**
Signed-off-by: robert pei <cspeiyb@gmail.com>

**- What I did**
Fix the issue as described in #31968 
**- How I did it**
0. if the filesize is smaller than the buffer then load the whole file into memory goto 6
1.Changed the reading the last n lines algorithm to:
2.seek the starting position for a buffer size block from EOF
3.count the number of "\n"'s
4.if number of eol is larger than n then return
5.if not then double the size of the buffer then go back to 1.
6.take the last n line of buffered file and return

My change implements the above Step#5.

**- How to verify it**
The following is my test result in my private test environment.
Before the change, it needs around 2 minutes to tail 100000 log lines.
[root@kube-node82 ~]#  time docker logs -t=true --tail=100000 f3b8d997959ad224  > test.log 2>&1
real	2m8.299s
user	0m0.379s
sys	0m0.382s

After the change, it only needs a couple of seconds as shown in the following.
[root@kube-node82 ~]#  time docker logs -t=true --tail=100000 80ef55861f5d3245  > test.log 2>&1

real	0m1.663s
user	0m0.320s
sys	0m0.371s

**- Description for the changelog**
Speed up tailing logs for json-file log driver

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by thaJeztah** [comment]
ping @unclejack @cpuguy83 PTAL

**Comment by haveatry** [comment]
@thaJeztah I found there were 2 failures in the first check item, But they seemed to be irrelevant to my code change, do I need to take any further action?

**Comment by tophj-ibm** [comment]
@haveatry no, I'm going to restart that job as I don't think it's related. The z and powerpc jobs were added recently, and with the increased scale we are still finding some flaky tests that they hit.

Opened #32267 for the flaky test.

**Comment by haveatry** [comment]
@vdemeester Do you have any comment on this fix? Do I need to take any further action?

**Comment by vdemeester** [comment]
@haveatry could you squash and sign off your commit ? 👼 🙏

**Comment by haveatry** [comment]
@vdemeester Just did it as your suggestion

**Comment by haveatry** [comment]
@vdemeester Do you have any comment on this fix? Do I need to take any further action?

File: pkg/tailfile/tailfile.go [hunk 1]
```diff
@@ -56,7 +58,10 @@ func TailFile(f io.ReadSeeker, n int) ([][]byte, error) {
 		if cnt > n {
 			break
 		}
-		block--
+		block = 2 * block
+		if block < maxBlock {
```
Comment (cpuguy83) [inline pkg/tailfile/tailfile.go]: Is this right? In this case `maxBlock` effectively looks like a min block.

**Comment by cpuguy83** [comment]
@haveatry It looks like the fix is to make the block size larger and larger?
Is tailing 10,000 lines a valid use-case? I should think at this point an external logger would be used.

File: pkg/tailfile/tailfile.go (hunk 1 above)
Comment (haveatry) [inline pkg/tailfile/tailfile.go]: If you feel confused about it, I could change the maxBlock to one positive value, and change the code.

**Comment by haveatry** [comment]
@cpuguy83 Thanks for your review. Yes, the fix is to make the block size larger and larger, but there is on limit. The largest block size is 1024*1024(blockSize * maxBlock). I am not sure whether tailing 10,000 line is valid or not, such operation is not forbidden at least.

**Comment by cpuguy83** [comment]
I'm -1 on this change.

Tailing a huge amount is abnormal and 1024B block size seems like a pretty good size for 10 lines (default size of the `tail` command).

**Comment by thaJeztah** [comment]
I also wonder if the intent in these situations is to have _all_ the logs, or a limited amount of logs (from a certain time-frame); if the latter, this PR may help for this; https://github.com/moby/moby/pull/32914 as it allows you to set an `--until` date/time to limit the logs that are returned

**Comment by thaJeztah** [comment]
Discussing this in the maintainers meeting, and we're not comfortable adding this enhancements; external log aggregation, or the `--since` / `--until` options are probably better alternatives for this.

I'm closing this PR, but thanks for contributing, and feel free to discuss 👍

**Pull request closed without merging**